evaluation/evaluator.go

```diff
@@ -301,11 +310,18 @@ func (e Evaluator) evaluate(identifier string, target *rest.Target, kind string)
 	if err != nil {
 		return rest.Variation{}, err
 	}
+	if e.postEvalChan != nil {
+		e.postEvalChan <- PostEvalData{
+			FeatureConfig: &flag,
+			Target:        target,
+			Variation:     &variation,
+		}
+	}
```

Comment:
I think this could potentially block forever if whatever is listening to `postEvalChan` isn't ready to receive. If we pass a context in to this function then we could do something like this so even if the write to the channel is blocked there's still a way to break out of this function.

```go
data :=  PostEvalData{
	FeatureConfig: &flag,
	Target:        target,
	Variation:     &variation,
}

// We wait until we can either write to the channel or the context has been cancelled.
select {
  case e.postEvalChan <- data:
  case <-ctx.Done():
      return rest.Variation{}, ctx.Err()
      }
}
```

It looks like this function gets called by `BoolVariation`, `StringVariation` so I don't know if it maybe makes sense to make them take a context and then pass it down to this function?

Reply:
I agree, but still it will block until receiver is ready, this will only solve when context is canceled for example on closing the client, right?

Reply:
Yea it will still block if the receiver isn't ready and can only be canceled by the client but it would at least mean we don't have a block of code that could block forever without giving the caller some way to stop it.

Reply:
Thinking out loud here and making some assumptions, but it looks like the purpose of the channel is to send evaluations to some caller that created the Evaluator using `NewEvaluator`? What if instead of using a channel we allowed them to pass in a function and then we call it instead of writing to a channel?

Something like this but with better naming of things than what I've given 😄 

```go
// 
type evalFn func (data PostEvalData)

func NewEvaluator(query Query, evalCallback evalFn) (*Evaluator, error) {
   ... code
}

func (e Evaluator) evaluate(identifier string, target *Target, kind string) (rest.Variation, error) {
   if e.query == nil {
		log.Errorf(ErrQueryProviderMissing.Error())
		return rest.Variation{}, ErrQueryProviderMissing
	}
	flag, err := e.query.GetFlag(identifier)
	if err != nil {
		return rest.Variation{}, err
	}

	if flag.Kind != kind {
		return rest.Variation{}, fmt.Errorf("%w, expected: %s, got: %s", ErrFlagKindMismatch, kind, flag.Kind)
	}

	if flag.Prerequisites != nil {
		prereq, err := e.checkPreRequisite(&flag, target)
		if err != nil || !prereq {
			return findVariation(flag.Variations, flag.OffVariation)
		}
	}
	variation, err := e.evaluateFlag(flag, target)
	if err != nil {
		return rest.Variation{}, err
	}
	if e.evalCallbackFn != nil {
		data := PostEvalData{
			FeatureConfig: &flag,
			Target:        target,
			Variation:     &variation,
		}

                e.evalCallback(data)
	}
	return variation, nil
}
```

Reply:
if you look prev commit it was done in that way but @davejohnston suggested to decouple using channels and go routine
[https://github.com/harness/ff-golang-server-sdk/pull/75/commits/ad78e24ffc37333123fae955004c3a8404e26a77#diff-4711ab68b7e3b590be90997cfb086cfe1e0d6456e76b35de90696561f6633962](url)

Reply:
@enver-bisevac I think there is a subtle difference in what James is proposing and the first revision.  And it makes sense.
   The Evaluator itself has the 'processor' rather than it being passed through from the each BoolVariation, StringVariation funcs
I'm fine with that, as it still separates the logic out - the caller of BoolVariation/IntVariation doesn't need to construct and pass in a 'processor' to each function.  Instead the evaluator decides what to do (based on how it was created)


```golang
func (e Evaluator) IntVariation(identifier string, target *Target, defaultValue int, processEvaluation ProcessEvaluation) int {

	variation, err := e.evaluate(identifier, target, "int", processEvaluation)
``` 

Instead it would look like this (assuming you follow James example above where the processor is a member of the evaluator).

```golang
func (e Evaluator) IntVariation(identifier string, target *Target, defaultValue int) int {
	variation, err := e.evaluate(identifier, target, "int")
```

Reply:
Ok, I agree with interface approach too, but I understand you woild like to do it in a manner of channel communication? So I will create a commit with interface as factory function